### PR TITLE
Fix directory-only glob for FileExplorer

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

Closes: #6686

Added a unit test too. The problem was that the previous glob-based implementation assumed directory paths always ended with `/`. When we switched to pathlib so that it could run on 3.8, we did not take that into account.

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

**Comment by abidlabs** [comment]
Otherwise, lgtm

**Comment by abidlabs** [comment]
Thanks @freddyaboulton! It does fix the empty file, but there's a different issue, which is that that there's no way to select a directory. Per @vchabaux's comment, previously you'd select the "empty file" to select the directory, but now there's no such file. 

I think the way it should work is that you should do something like this:

```py
import gradio as gr


demo = gr.Interface(
    lambda x:x,
    gr.FileExplorer(glob="**/", file_count="directory"),
    "textbox"
)

demo.launch()
```

and then you should be able to select a directory to pass into your prediction function. However, that doesn't seem to work -- nothing gets passed into the function even if you select a directory.

**Comment by abidlabs** [comment]
Taking another look

**Comment by abidlabs** [comment]
@freddyaboulton I'm not able to select folders. What's the code for your demo? Here's what I tried:

```py
import gradio as gr


demo = gr.Interface(
    lambda x:x,
    gr.FileExplorer(glob="**/", file_count="single"),
    "textbox"
)

demo.launch()
```

**Comment by abidlabs** [comment]
Hmm that behavior doesn't make sense to me -- so `single` doesn't allow you to select directories, but `multiple` does?

**Comment by vchabaux** [comment]
Initially my usage was to select a directory on the machine where python runs (and only one) as the directory for my output. I think this usage is quite common, and it would be sad not to be able to use FileExplorer for this.

**Comment by freddyaboulton** [comment]
@pngwn @abidlabs i think this is finally working as intended. If file_count=`single`, clicking on a directory won't automatically select all its children.

**Comment by abidlabs** [comment]
@freddyaboulton here's my proposal to address the point (2) above: we add an `include_dirs` parameter in `FileExplorer`, which is `False` by default. Essentially it allows users to select directories in `FileExplorer`, but it has two effects: one effect in the UI, and one effect in the data that is passed into the user function

We can consider the cases `file_count=="single"` and `file_count=="multiple"` separately:

* If `file_count=="single"` and `include_dirs is True`, then one is able to select a directory, and it only selects that particular directory. Naturally, it passes the selected directory into the user function. Note that if `include_dirs is False` and `file_count=="single"`, we keep the current behavior, which disables directory selection altogether. 

 
* If `file_count="multiple"` and `include_dirs is True`, then I'd suggest that selecting a directory *does not* automatically select its children. Multiple directories and/or files are selected, and they are all passed into the user function individually. If `include_dirs is False` we do not pass directories into the user function. Selecting a directory has the effect of selecting of its all children files and only those files are passed into the user function.

tldr: setting `include_dirs` to `True` allows users to select directories individually in the `FileExplorer` and those are sent to the user function. Keeping `include_dirs` as `False` guarantees that only files are sent to the user function, no directories.

**Comment by freddyaboulton** [comment]
@abidlabs @pngwn I implemented what we discussed earlier - including the directory in the file tree with a new directory icon and the value ".".

The default glob pattern excludes directories (that's behavior on `main`). You can glob for everything with `glob="**/*"` and you can ignore directories by using the default glob or with `ignore_glob="**/"`. You can include only directories with `glob="**/"`.

Here is a gif showing what it looks like now:

file_count_multiple

File: js/fileexplorer/icons/light-folder.svg
Comment (abidlabs) [inline js/fileexplorer/icons/light-folder.svg:1]: Nice where's this svg from?

**Comment by abidlabs** [comment]
Fantastic solution @freddyaboulton! Tested it out and working well for me

**Comment by abidlabs** [comment]
I think @vchabaux will be a fan too

**Review by abidlabs — APPROVED** [review]

**Comment by vchabaux** [comment]
Awesome solution, thanks a lot @freddyaboulton  !
